File: frontend/lib/utils/date.dart
import 'package:holiday_jp/holiday_jp.dart' as holiday_jp;

String getJPDayOfWeekString(int dayOfWeek) {
  switch (dayOfWeek) {
    case 1:
      return '月';
    case 2:
      return '火';
    case 3:
      return '水';
    case 4:
      return '木';
    case 5:
      return '金';
    case 6:
      return '土';
    case 7:
      return '日';
    default:
      return '';
  }
}

bool isJPWeekday(DateTime date) {
  final isHoliday = holiday_jp.isHoliday(date);
  return date.weekday <= DateTime.friday && !isHoliday;
}

bool isSameDate(DateTime date1, DateTime date2) {
  return date1.year == date2.year &&
      date1.month == date2.month &&
      date1.day == date2.day;
}

/// 開催中なら開催中の日付を、開催中でなければ次の開催日を返す
List<DateTime> getOngoingOrComingWeekdays() {
  final today = DateTime.now();

  // 今週の月曜日から金曜日のうち、平日の日付を取得
  final thisMonday = today.subtract(Duration(days: today.weekday - 1));
  List<DateTime> thisWeekdays = [];
  for (int i = 0; i < 5; i++) {
    final date = thisMonday.add(Duration(days: i));
    if (isJPWeekday(date)) {
      thisWeekdays.add(date);
    }
  }
  // print("This weekdays: $thisWeekdays");

  // 今週の開催がまだ終わっていなければ、今週の開催日を返す
  if (thisWeekdays.isNotEmpty &&
      (today.isBefore(thisWeekdays.last) ||
          today.isAtSameMomentAs(thisWeekdays.last))) {
    return thisWeekdays;
  }

  // 今週の開催が終わっている場合は、次回の開催日を返す
  final nextMonday = thisMonday.add(Duration(days: 7));
  List<DateTime> nextWeekdays = [];
  for (int i = 0; i < 5; i++) {
    final date = nextMonday.add(Duration(days: i));
    if (isJPWeekday(date)) {
      nextWeekdays.add(date);
    }
  }
  // print("Next weekdays: $nextWeekdays");
  return nextWeekdays;
}

File: backend/functions/src/utils/date.ts

import * as holiday_jp from '@holiday-jp/holiday_jp';
import { logger } from 'firebase-functions';

const jpDaysOfWeek = {
    0: "日",
    1: "月",
    2: "火",
    3: "水",
    4: "木",
    5: "金",
    6: "土",
}


// Get the current date in Japan timezone
function getCurrentJpDate(): Date {
    const now = new Date();
    const jpDateString = now.toLocaleString('ja-JP', { timeZone: 'Asia/Tokyo' });
    const jpDate = new Date(jpDateString);
    return jpDate;
}

// Get the date string in the format YYYY-MM-DD
function getDateKey(date: Date): string {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
}


function isJpWeekday(date: Date): boolean {
    const day = date.getDay();
    if (day === 0 || day === 6) {
        return false;
    }
    if (holiday_jp.isHoliday(date)) {
        return false;
    }
    return true;
}

function getJpDateString(date: Date, withDayOfWeek: boolean = true): string {
    if (withDayOfWeek) {
        return `${date.getMonth() + 1}/${date.getDate()} (${getJpDayOfWeek(date)})`;
    }
    return `${date.getMonth() + 1}/${date.getDate()}`;
}

function getJpDayOfWeek(date: Date): string {
    return jpDaysOfWeek[date.getDay()];
}

function getOngoingOrLastWeekdays(): Date[] {
    /*
    開催中なら開催中の日付を，開催中でなければ前回の開催日を返す
    (結果の確認などに使う)
    */
    const today = getCurrentJpDate();
    let todayDayOfWeek = today.getDay();

    // 今週の月曜日から金曜日のうち、平日の日付を取得
    const thisMonday = new Date(today.getFullYear(), today.getMonth(), today.getDate() - todayDayOfWeek + 1);
    const thisWeekdays = [];
    for (let i = 0; i < 5; i++) {
        const date = new Date(thisMonday);
        date.setDate(thisMonday.getDate() + i);
        if (isJpWeekday(date)) {
            thisWeekdays.push(date);
        }
    }
    logger.log("This weekdays:", thisWeekdays);

    // 今週の開催がもう始まっていれば、今週の開催日を返す
    if (thisWeekdays.length != 0 && thisWeekdays[0] <= today) {
        return thisWeekdays;
    }

    // まだ開催されていない場合は、前回の開催日を返す
    const lastMonday = new Date(thisMonday.getFullYear(), thisMonday.getMonth(), thisMonday.getDate() - 7);
    const lastWeekdays = [];
    for (let i = 0; i < 5; i++) {
        const date = new Date(lastMonday);
        date.setDate(lastMonday.getDate() + i);
        if (isJpWeekday(date)) {
            lastWeekdays.push(date);
        }
    }
    logger.log("Last weekdays:", lastWeekdays);
    return lastWeekdays;
}



export { getCurrentJpDate, getDateKey, isJpWeekday, getJpDateString, getJpDayOfWeek, getOngoingOrLastWeekdays };